store.go

```diff
@@ -0,0 +1,254 @@
+//  - Block-data is transparently stored zstd-compressed where sensible
+//  - There is ability to efficiently record and query DAG-forming block relationships directly in the database
+//  - If configured with an instance namespace, keeps a log of recently-accessed blocks, which is then used to bulk-load blocks into the LRU cache on cold-starts
+//  - Has a mode tracking every Read/Write with millisecond precision
+type PgBlockstore struct {
+	isWritable                 bool
+	cacheInactiveBeforeRead    bool
+	preloadRecents             bool
+	parseBlockLinks            bool
+	prefetchDagLayersOnDbRead  int32
+	lruSizeBytes               int64
+	instanceNamespace          string
+	additionalPreloadNamespace string
+	lru                        cidkeyedlru.CidKeyedLRU
+	dbPool                     *pgxpool.Pool
+	accessLogsMu               sync.Mutex
+	accessLogsRecent           map[int64]struct{}
+	accessLogsDetailed         map[accessUnit]int64
+	limiterLogsDetailedWrite   chan struct{}
+	limiterSetLastAccess       chan struct{}
+	limiterBlockProcessing     chan struct{}
+	shutdownSemaphore          chan struct{}
+
+	// all manipulated concurrently via atomic.Store/Load
+	firstReadPerformed   *int32
+	lastFlushEpoch       *int64
+	linearSyncEventCount *int64
+}
+
+// Close releases the database connection and purges the cache. The store can
+// not be used again after it has been closed.
+func (dbbs *PgBlockstore) Close() error {
+	if dbbs == nil {
+		return nil
+	}
+	log.Infof("shut down of %T(%p) begins", dbbs, dbbs)
+	close(dbbs.shutdownSemaphore)
+	dbbs.dbPool.Close()
+	dbbs.lru.Purge()
+	log.Infof("shut down of %T(%p) completed", dbbs, dbbs)
+	return nil
+}
+
+// PgxPool returns a reference to the underlying pgx.Pool
+func (dbbs *PgBlockstore) PgxPool() *pgxpool.Pool { return dbbs.dbPool }
+
+// InstanceNamespace returns the current configuration,
+// "" means no namespace has been set.
+func (dbbs *PgBlockstore) InstanceNamespace() string { return dbbs.instanceNamespace }
+
+// IsWritable indicates that the store was configured to be Writable
+// and we are connected to a RDBMS in a way permitting writes.
+func (dbbs *PgBlockstore) IsWritable() bool { return dbbs.isWritable }
+
+// maybeLogUnexpectedErrorf is a helper to log internal unexpected storage
+// errors before sending them up the chain where possible (e.g. the Block
+// interface has no provisions for error reporting at all)
+func (dbbs *PgBlockstore) maybeLogUnexpectedErrorf(format string, a ...interface{}) {
+	select {
+	case <-dbbs.shutdownSemaphore:
+		// if we are shutting down - nothing unexpected about it
+	default:
+		log.Errorf("UNEXPECTED: "+format, a...)
+	}
+}
+
+//
+// Start of blockstore.Blockstore implementation
+//
+
+// DeleteBlock is NOT IMPLEMENTED by this store - the storage is strictly
+// append-only. Always returns an error upon invocation.
+func (NIM *PgBlockstore) DeleteBlock(cid.Cid) (err error) {
```

Comment:
Driveby comment: with recent changes in lotus (if this is trying to match the lotus blockstore interface(?)) this will need to be extended to meet the new interface: https://github.com/filecoin-project/lotus/blob/4c05ec28ba45e6947118c5a0bc836f312442d0e9/blockstore/blockstore.go#L20-L21

Reply:
Yeah definitely,haven't had a chance to upgrade past 1.5.0 yet
Keep the comments coming - I am slowly working through them ( this PR will be split into 3 as a result of some other discussions, but all the same: I do read )

Reply:
Addressed in https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/6f5650b22ffa01ff1d2c021916d85d4e7e91e072#diff-ed648181f98484bd12541509e0ae7b5ad1d1de7674ab1721814b47ddd5c95de4R124-R131